_posts/2023-05-06-reverse-engineering-gba.markdown

```diff
@@ -0,0 +1,1064 @@
+
+The 0x64 in `r10` came from `r2`, but there are no more references to `r2` between here and the start of the function. This means that `r2` is the third parameter of `FUN_08077c44`, and the 0x64 came from the code that called `FUN_08077c44`. The next step is to find the caller of this function.
+
+Scroll to the bottom of `FUN_08077cc44`, and you'll find the `bx` instruction at 0x8077dd6 marking the end of the function. Make a breakpoint in mGBA at 0x8077dd6, then continue program execution to hit the breakpoint. Note that before hitting the breakpoint, you'll hit the watchpoint for player HP again; this is where the player's HP is capped at their max HP.
+
+>![](/assets/img/reverse-engineering/oran-berry-end-function.png)<br>
+>Hitting the breakpoint at the end of `FUN_08077cc4`
+
+The breakpoint shows that `r0` is 0x0804838D. As the code is about to return from `FUN_08077cc4`, this is where the function will return to, and also where it was called. However, if you go to address 0x804838D in Ghidra, this section is not marked as assembly code.
+>![](/assets/img/reverse-engineering/oran-berry-missing-assembly.png)<br>
+>Missing assembly code at 0x804838D
+
+Sometimes Ghidra doesn't auto-detect certain blocks of code as assembly. Select line 0x8048364 (directly after the last function that is already disassembled) and press F12 to disassemble this block of code as THUMB assembly. Ghidra will disassemble this line and also follow the assembly branches and function calls to disassemble subsequent lines of assembly, including line 0x804838D.
+
+>![](/assets/img/reverse-engineering/oran-berry-fixed-assembly.png)<br>
+>0x804838D and surrounding addresses after manually disassembling it
+
+Now the call to `FUN_08077cc44` is visible at address 0x8048388. We were looking for where `r2` was assigned to 0x64, so look at the lines prior to the function call. We can see this set of instructions starting at 0x8048378.
+```
+ldr r2,[DAT_08048394]
+mov r3,#0x0
+ldrsh r2,[r2,r3]
+```
+This loads an address into `r2` from data value `DAT_08048394`, then loads the value at the address into `r2`. Look at address 0x48394 to find that the value of `DAT_08048394` is 0x080F4FB6. This is within ROM, which is promising; ROM cannot normally be written to, so chances are that this is the source of the value 0x64.
+
+With this address in mind, go back to mGBA and go to address 0x080F4FB6 in the memory viewer. You will find the value 0x64 at the address, confirming that this is the location of the Oran Berry's HP heal amount.
+>![](/assets/img/reverse-engineering/oran-berry-found.png)<br>
+>The location of the Oran Berry's heal amount
+
+To test the address we found for the heal amount, reload your save state, change the 0x64 at address 0x080F4FB6 to 0x0A (10 in decimal) in the memory viewer, and eat the Oran Berry again to see how much it heals. You may wish to delete the watchpoint and breakpoint from earlier to avoid hitting them again now that we are finished with them.
```

Comment:
```suggestion
To test the address we found for the heal amount, reload your save state, change the 0x64 at address 0x080F4FB6 to 0x0A ("0x0A" is 10 in decimal) in the memory viewer, and eat the Oran Berry again to see how much it heals. You may wish to delete the watchpoint and breakpoint from earlier to avoid hitting them again now that we are finished with them.
```